src/main/java/net/minecraftforge/event/ModMismatchEvent.java

```diff
@@ -0,0 +1,184 @@
+     * @param modId The mod to fetch previous version for.
+     * @return The previously known mod version, or {@link Optional#empty()} if unknown/not found.
+     */
+    @Nullable
+    public ArtifactVersion getPreviousVersion(String modId)
+    {
+        if(this.versionDifferences.containsKey(modId))
+            return this.versionDifferences.get(modId).oldVersion();
+
+        return null;
+    }
+
+    @Nullable
+    public ArtifactVersion getCurrentVersion(String modid)
+    {
+        if(this.versionDifferences.containsKey(modid))
+            return this.versionDifferences.get(modid).newVersion();
+
+        return null;
+    }
+
+    /**
+     * Marks the mod version mismatch as having been resolved safely by the current mod.
+     */
+    public void markResolved(String modId)
+    {
+        final var resolvedBy = ModLoadingContext.get().getActiveContainer();
+        resolved.putIfAbsent(modId, resolvedBy);
+    }
+
+    /**
+     * Fetches the status of a mod mismatch handling state.
+     */
+    public boolean wasResolved(String modId)
+    {
+        return this.resolved.containsKey(modId);
+    }
+
+    public Optional<MismatchedVersionInfo> getVersionDifference(String modid)
+    {
+        return Optional.ofNullable(this.versionDifferences.get(modid));
+    }
+
+    public Optional<ModContainer> getResolver(String modid)
+    {
+        return Optional.ofNullable(this.resolved.get(modid));
+    }
+
+    public boolean anyUnresolved()
+    {
+        return resolved.size() < versionDifferences.size();
+    }
+
+    public Stream<MismatchResolutionResult> getUnresolved()
+    {
+        return versionDifferences.keySet().stream()
+                .filter(modid -> !resolved.containsKey(modid))
+                .map(unresolved -> new MismatchResolutionResult(unresolved, versionDifferences.get(unresolved), null))
+                .sorted(Comparator.comparing(MismatchResolutionResult::modid));
+    }
+
+    public boolean anyResolved()
+    {
+        return !resolved.isEmpty();
+    }
+
+    public Stream<MismatchResolutionResult> getResolved()
+    {
+        return resolved.keySet().stream()
+                .map(modid -> new MismatchResolutionResult(modid, versionDifferences.get(modid), resolved.get(modid)))
+                .sorted(Comparator.comparing(MismatchResolutionResult::modid));
+    }
+
+    public record MismatchResolutionResult(String modid, MismatchedVersionInfo versionDifference, @Nullable ModContainer resolver) {
+        public boolean wasSelfResolved()
+        {
+            return resolver != null && resolver.getModId().equals(modid);
+        }
+
+    }
+
+    public record MismatchedVersionInfo(ArtifactVersion oldVersion, @Nullable ArtifactVersion newVersion)
+    {
+        public boolean isMissing()
+        {
+            return newVersion == null;
+        }
+
+        public boolean wasUpgrade()
+        {
+            if(newVersion == null) return false;
```

Comment:
Formatting: space between `if` keyword and parenthesis.
```suggestion
            if (newVersion == null) return false;
```